proposals/NNNN-package-manager-product-definitions.md

```diff
@@ -189,105 +167,92 @@ As with targets, there is no semantic significance to the order of the products
+### Not adding product definitions
 
-## Open questions
+Instead of product definitions, fine-grained dependencies could be introduced by allowing targets to be marked as public or private to the package.
 
-1. Should tests be considered to be products?
+_Advantage_
 
-   Pro:  Since tests are artifacts that can be produced during a build, just like products are, it may make sense for users to be able to express opinions about them in the same manner as for products.
-   
-   Con:  Tests are conceptually different from products in their intended use, and it is much less likely that users will have specific opinions about the details of how tests are built.  It seems conceptually cleaner to treat tests as a different kind of artifact than products.
-   
-2. Should there be any inference of products at all?
+It would avoid the need to introduce new concepts.
+
+_Disadvantage_
+
+It would not provide any way for a package author to control the type and characteristics of the various artifacts.  Relying on the implied products that result from the inference rules is not likely to be scalable in the long term as we introduce new kinds of product types.
+
+### Inferring implicit products
+
+An obvious alternative to the proposal would be to keep the inference rules even in v4 of the PackageDescription API.
 
-   Pro:  It can be much more convenient to not have to declare products, and to be able to add new products just by adding or renaming files and directories.
+_Advantage_
 
-   Con:  The very fact that it is so easy to change the set of products without modifying the package manifest can lead to unexpected behavior due to seemingly unrelated changes (e.g. creating a file called "main.swift" in a module changes that module from a library to an executable).
+It can be a lot more convenient to not have to declare products, and to be able to add new products just by adding or renaming files and directories.  This is particularly true for small, simple packages.
 
-3. Should target dependencies on products be separate from dependencies on other targets?
+_Disadvantages_
 
-   Pro:  Targets and products are different concepts, and it is possible for a target and a product to have the same name.  This can lead to ambiguity.
+The very fact that it is so easy to change the set of products without modifying the package manifest can lead to unexpected behavior due to seemingly unrelated changes (e.g. creating a file called `main.swift` in a module changes that module from a library to an executable).
+
+Also, as packages become more complex and new conceptual parts on which clients can depend are introduced, the interaction of implicit rules with the explicit product definitions can become very complicated.
+
+We plan to provide some of the convenience through tooling.  For example, an IDE (or `swift package` itself on the command line) can offer to add product definitions when it notices certain types of changes to the structure of the package.
```

Comment:
I'd add a comment in this section that having declared types lets us present better diagnostics and helps us "guide" people to learning the expected conventions.